folly/io/async/test/AsyncSSLSocketTest.cpp

```diff
@@ -1525,6 +1525,79 @@ static int kRSAEvbExIndex = -1;
 static int kRSASocketExIndex = -1;
 static constexpr StringPiece kEngineId = "AsyncSSLSocketTest";
 
+int customRsaPubDec(int flen, const unsigned char *from, unsigned char *to, RSA *rsa,
+                                          int padding) {
+
+  ScopedEventBaseThread jobEvbThread;
+  EventBase* asyncJobEvb = jobEvbThread.getEventBase();
+  AsyncSSLSocket* socket = reinterpret_cast<AsyncSSLSocket*>(
+      RSA_get_ex_data(rsa, kRSASocketExIndex));
+  ASYNC_JOB* job = ASYNC_get_current_job();
+  if (job == nullptr) {
+    throw std::runtime_error("Expected call in job context");
+  }
+
+  ASYNC_WAIT_CTX* waitctx = ASYNC_get_wait_ctx(job);
+  OSSL_ASYNC_FD pipefds[2] = {0, 0};
+  ASYNC_WAIT_CTX_clear_fd(waitctx, kEngineId.data());
+  makeNonBlockingPipe(pipefds);
+  if (!ASYNC_WAIT_CTX_set_wait_fd(
+          waitctx, kEngineId.data(), pipefds[0], nullptr, nullptr)) {
+    throw std::runtime_error("Cannot set wait fd");
+  }
+
+  int ret = 0;
+  int* retptr = &ret;
+
+  auto hand = folly::NetworkSocket::native_handle_type(pipefds[1]);
+  auto asyncPipeWriter = folly::AsyncPipeWriter::newWriter(
+      asyncJobEvb, folly::NetworkSocket(hand));
+  asyncJobEvb->runInEventBaseThread([retptr = retptr,
+                                     flen = flen,
+                                     from = from,
+                                     to = to,
+                                     padding = padding,
+                                     actualRSA = rsa,
+                                     writer = std::move(asyncPipeWriter),
+                                     socket = socket]() {
+    LOG(INFO) << "Running job";
+    if (socket) {
+      LOG(INFO) << "Got a socket passed in, closing it...";
+      socket->closeNow();
+    }
+    *retptr = RSA_meth_get_pub_dec(RSA_PKCS1_OpenSSL())(
+        flen, from, to, actualRSA, padding);
+    LOG(INFO) << "Finished job, writing to pipe";
+    uint8_t byte = *retptr > 0 ? 1 : 0;
+    writer->write(nullptr, &byte, 1);
+  });
+
+  LOG(INFO) << "About to pause job";
+  ASYNC_pause_job();
+  LOG(INFO) << "Resumed job with ret: " << ret;
+  return ret;
+}
+
+int verifyCb(X509_STORE_CTX *ctx, void *arg) {
+  int ok = 1;
+  X509 *cert = X509_STORE_CTX_get0_cert(ctx);
+  CHECK(cert);
+
+  ssl::EvpPkeyUniquePtr publicEvpPkey(X509_get_pubkey(cert));
+  EVP_PKEY *pkey = publicEvpPkey.get();
+  RSA* rsa = EVP_PKEY_get1_RSA(publicEvpPkey.get());
```

Comment:
I think this is a leak (openssl _get1 increments (adds by 1) the reference count of the undelrying object), and the reason why tests aren't catching this is because we disable this test when running under ASAN mode.

Reply:
See line 1762 for where we disable